charm/src/charm.py

```diff
@@ -107,8 +113,50 @@ def _on_ingress_ready(self, _: ops.HookEvent) -> None:
     def reconcile(self) -> None:
         """Reconcile all services."""
         self._add_netbox_rq()
+        self._add_command_to_cron(
+            CRON_EVERY_5_MINUTES, "syncdatasource", "/bin/python3 manage.py syncdatasource --all"
+        )
+        self._add_command_to_cron(
+            CRON_AT_MIDNIGHT, "housekeeping", "/bin/python3 manage.py housekeeping"
+        )
         super().reconcile()
 
+    def _add_command_to_cron(self, scheduling: str, name: str, command: str) -> None:
+        """Add a command that will be run with cron.
+
+        Args:
+            scheduling: scheduling following cron format.
+            name: name for the pebble service to run. Should be compliant with pebble
+               service names.
+            command: command to execute.
+        """
+        container = self.workload()
+        if not container.can_connect():
+            return
+        layer: ops.pebble.LayerDict = {
+            "services": {
+                name: {
+                    "override": "replace",
+                    "summary": f"NetBox {name}",
+                    "startup": "disabled",
+                    "on-success": "ignore",
+                    # the command should last at least 1 second so pebble thinks
+                    # everything is correct.
+                    "command": f"run_after_1s {command}",
+                    "working-dir": str(self._BASE_DIR / "app"),
+                    "environment": self.gen_env(),
+                    "user": "_daemon_",
+                }
+            }
+        }
+        container.add_layer(name, layer, combine=True)
+        container.push(
+            f"/etc/cron.d/{name}",
+            f"{scheduling} root "
+            + f"PEBBLE_SOCKET=/charm/container/pebble.socket pebble start {name}\n",
```

Comment:
I think this is super clever to use pebble services to store all command parameters, like environmental variables. However, personally, I think this might be a little bit outside the recommend application of the pebble service.

One alternative method I can think of is that crontab supports defining environment variables within it. So, one could configure the crontab file as follows (untested):

```
AWS_ENDPOINT_URL=https://foobar.test
POSTGRESQL_DB_CONNECT_STRING=postgresql://foobar.test

0 0 * * * _daemon_ cd /django/app && /bin/python3 manage.py syncdatasource --all
*/5 * * * * _daemon_ cd /django/app && /bin/python3 manage.py housekeeping
```

The main disadvantage here is it doesn't direct output to pebble logs.

Another solution I propose is to create a straightforward Python script to run the housekeeping tasks repeatedly.

```python
#!/bin/python3
import argparse
import math
import random
import subprocess
import time

parser = argparse.ArgumentParser(description="run command repeatedly")
parser.add_argument("--every", type=int, help="interval in seconds", required=True)
parser.add_argument("command", nargs=argparse.REMAINDER)

args = parser.parse_args()
interval = args.every
command = args.command

while True:
    next_time = math.ceil(time.time() / interval) * interval
    time.sleep(next_time - time.time() + interval * random.random() * 0.1)  # add 10% random delay
    subprocess.run(command)
```

And the pebble service configuration can be a normal service definition:

```yaml
startup: enabled
command: /bin/repeat.py --every 300 /bin/python3 manage.py syncdatasource --all
environments: {...}
user: _daemon_
working-dir: /django/app
```

Of cause I think keep using the service is perfectly fine, what do you think?

Reply:
You are right, I am "stretching" a bit the definition of a pebble service to get the advantage of env variables.

Actually, seeing your ideas, I found another option, I could also run pebble exec, that does not stretch the definition of a service so much, and inherit env variables from another service, like:
```
pebble exec -w="/django/app" --context=django -- /bin/python3 manage.py syncdatasource --all
```

Reply:
@benhoyt would you recommend any of the proposed solutions? Is the current option one to disregard?

Reply:
I'd *strongly* recommend using `pebble exec` for this, as it was designed for basically this exact use case (especially with the `--context` parameter, which we added relatively recently to simplify the env vars handling for this case).

`pebble logs` is only for service logs, so yes, for `pebble exec` you'd need to send the logs somewhere. However, the logs will come out stdout/stderr, so you can just wrap the exec command or redirect it if you want to send those somewhere else (or use a nicer cron runner that takes care of that).

Similar for retries, you'd need to wrap the command that you're exec-ing to do retries (again, a "better" cron runner could take care of that).

Reply:
I was just reading the pebble document, and I came up with something.

What if we set the `backoff-factor` to `1.0`? Does that mean this will become a very rudimentary timer service? For example, the following service will restart every 5 minutes.

```yaml
startup: enabled
command: /bin/python3 manage.py syncdatasource --all
environments: {...}
user: _daemon_
working-dir: /django/app
on-success: restart
on-failure: restart
backoff-delay: 5m
backoff-factor: 1.0
backoff-limit: 5m
```

Too much?

Reply:
Stuff like this is clever, but not really what services (and auto-restarts) were designed for. If you want to hack something like this, a health check on an interval might be a better option. :-) But once again, I recommend using a proper cron tool and/or `pebble exec`.

Reply:
Thanks both of you for the (cool) ideas! As suggested, I will replace the service for exec.